lib/gql_serializer.rb

```diff
@@ -46,55 +46,71 @@ def self.query_include(model, hasharray)
 
     if record.nil?
       return nil
     end
 
+
     if record.respond_to? :map
       return record.map do |r|
-        self.serialize(r, hasharray, options)
+        self.serialize(r, hasharray, options, instructions)
       end
     end
 
-    hash = {}
-    model = record.class
-    all_relations = model.reflections.keys
+    id = "#{record.class}, #{hasharray}"
+    instruction = instructions[id]
+    if (!instruction)
+      instruction = {klass: record.class, hasharray: hasharray, relations: [], attributes: []}
+      instructions[id] = instruction
 
-    relations = hasharray.filter do |e|
-      next true if !e.is_a?(String)
 
-      key, alias_key = e.split(':')
-      all_relations.include?(key)
-    end
+      model = record.class
+      all_relations = model.reflections.keys
 
-    attributes = hasharray - relations
-    attributes = model.attribute_names if attributes.empty?
+      relations = hasharray.filter do |e|
+        next true if !e.is_a?(String)
 
-    attributes.each do |e|
-      key, alias_key = e.split(':')
-      alias_key = apply_case(alias_key || key, options[:case])
+        key, alias_key = e.split(':')
 
-      hash[alias_key] = coerce_value(record.public_send(key))
-    end
+        all_relations.include?(key)
+      end
 
-    relations.each do |e|
-      if e.is_a? String
+      attributes = hasharray - relations
+      attributes = model.attribute_names if attributes.empty?
+
+      attributes.each do |e|
         key, alias_key = e.split(':')
         alias_key = apply_case(alias_key || key, options[:case])
 
-        rel_records = record.public_send(key)
-        hash[alias_key] = self.serialize(rel_records, [], options)
-      else
-        key, alias_key = e.keys.first.split(':')
-        alias_key = apply_case(alias_key || key, options[:case])
+        instruction[:attributes].push({key: key, alias_key: alias_key})
+      end
+
+      relations.each do |e|
+        if e.is_a? String
+          key, alias_key = e.split(':')
+          alias_key = apply_case(alias_key || key, options[:case])
 
-        rel_records = record.public_send(key)
-        hash[alias_key] = self.serialize(rel_records, e.values.first, options)
+          instruction[:relations].push({key: key, alias_key: alias_key, hasharray: []})
+        else
+          key, alias_key = e.keys.first.split(':')
+          alias_key = apply_case(alias_key || key, options[:case])
+
+          instruction[:relations].push({key: key, alias_key: alias_key, hasharray: e.values.first})
+        end
       end
     end
 
+    hash = {}
+    instruction[:attributes].each do |attribute|
+      hash[attribute[:alias_key]] = coerce_value(record.public_send(attribute[:key]))
+    end
+    instruction[:relations].each do |relation|
+      relation_records = record.public_send(relation[:key])
+      hash[relation[:alias_key]] = self.serialize(relation_records, relation[:hasharray], options, instructions)
+    end
+
     hash
   end
 
```

Comment:
I'm thinking we should have some basic tests for the functionality if we're going to be making changes like this.

Reply:
We do have tests.

Reply:
Check out https://github.com/TheDro/gql_serializer/blob/master/spec/gql_serializer_spec.rb

Reply:
Oh that's great, thank you!